nameserver/server_cache_test.go

```diff
@@ -11,6 +11,247 @@ import (
 	wt "github.com/weaveworks/weave/testing"
 )
 
+// Check that AddRecord/DeleteRecord/... in the Zone database lead to cache invalidations
+func TestServerDbCacheInvalidation(t *testing.T) {
+	const (
+		containerID = "somecontainer"
+		testName1   = "first.weave.local."
+		testName2   = "second.weave.local."
+	)
+
+	InitDefaultLogging(testing.Verbose())
+	Info.Println("TestServerDbCacheInvalidation starting")
+
+	clk := newMockedClock()
+
+	Debug.Printf("Creating mocked mDNS client and server")
+	mdnsServer1 := newMockedMDNSServerWithRecord(Record{testName1, net.ParseIP("10.2.2.9"), 0, 0, 0})
+	mdnsCli1 := newMockedMDNSClient([]*mockedMDNSServer{mdnsServer1})
+
+	Debug.Printf("Creating zone database with the mocked mDNS client and server")
+	zoneConfig := ZoneConfig{
+		MDNSServer: mdnsServer1,
+		MDNSClient: mdnsCli1,
+		Clock:      clk,
+	}
+	zone, err := NewZoneDb(zoneConfig)
+	wt.AssertNoErr(t, err)
+	err = zone.Start()
+	wt.AssertNoErr(t, err)
+	defer zone.Stop()
+
+	Debug.Printf("Creating a cache")
+	cache, err := NewCache(1024, clk)
+	wt.AssertNoErr(t, err)
+
+	fallbackHandler := func(w dns.ResponseWriter, req *dns.Msg) {
+		m := new(dns.Msg)
+		m.SetReply(req)
+		if len(req.Question) == 1 {
+			m.Rcode = dns.RcodeNameError
+		}
+		w.WriteMsg(m)
+	}
+
+	// Run another DNS server for fallback
+	fallback, err := newMockedFallback(fallbackHandler, nil)
+	wt.AssertNoErr(t, err)
+	fallback.Start()
+	defer fallback.Stop()
+
+	Debug.Printf("Creating a real DNS server with a mocked cache")
+	srv, err := NewDNSServer(DNSServerConfig{
+		Zone:              zone,
+		Cache:             cache,
+		Clock:             clk,
+		ListenReadTimeout: testSocketTimeout,
+		UpstreamCfg:       fallback.CliConfig,
+		MaxAnswers:        4,
+	})
+	wt.AssertNoErr(t, err)
+	defer srv.Stop()
+	go srv.Start()
```

Comment:
It doesn't make a huge amount of difference, but `go ...Start()` appears before `defer ...Stop()` everywhere else...